File: src/python/4_BehavioralPatterns/9_Strategy/exercise.py
import math
import cmath
from unittest import TestCase
from abc import ABC


class DiscriminantStrategy(ABC):
    def calculate_discriminant(self, a, b, c):
        pass


class OrdinaryDiscriminantStrategy(DiscriminantStrategy):
    def calculate_discriminant(self, a, b, c):
        return b*b - 4*a*c


class RealDiscriminantStrategy(DiscriminantStrategy):
    def calculate_discriminant(self, a, b, c):
        result = b*b-4*a*c
        return result if result >= 0 else float('nan')


class QuadraticEquationSolver:
    def __init__(self, strategy):
        self.strategy = strategy

    def solve(self, a, b, c):
        """ Returns a pair of complex (!) values """
        disc = complex(self.strategy.calculate_discriminant(a, b, c), 0)
        root_disc = cmath.sqrt(disc)
        return (
            (-b + root_disc) / (2 * a),
            (-b - root_disc) / (2 * a)
        )

class Evaluate(TestCase):
    def test_positive_ordinary(self):
        strategy = OrdinaryDiscriminantStrategy()
        solver = QuadraticEquationSolver(strategy)
        results = solver.solve(1, 10, 16)
        self.assertEqual(complex(-2, 0), results[0])
        self.assertEqual(complex(-8, 0), results[1])

    def test_positive_real(self):
        strategy = RealDiscriminantStrategy()
        solver = QuadraticEquationSolver(strategy)
        results = solver.solve(1, 10, 16)
        self.assertEqual(complex(-2, 0), results[0])
        self.assertEqual(complex(-8, 0), results[1])

    def test_negative_ordinary(self):
        strategy = OrdinaryDiscriminantStrategy()
        solver = QuadraticEquationSolver(strategy)
        results = solver.solve(1, 4, 5)
        self.assertEqual(complex(-2, 1), results[0])
        self.assertEqual(complex(-2, -1), results[1])

    def test_negative_real(self):
        strategy = RealDiscriminantStrategy()
        solver = QuadraticEquationSolver(strategy)
        results = solver.solve(1, 4, 5)
        self.assertTrue(math.isnan(results[0].real))
        self.assertTrue(math.isnan(results[1].real))
        self.assertTrue(math.isnan(results[0].imag))
        self.assertTrue(math.isnan(results[1].imag))

File: src/swift/4_BehavioralPatterns/10_Strategy/Exercise.swift
import Foundation
import XCTest

protocol DiscriminantStrategy
{
  func calculateDiscriminant(_ a: Double, _ b: Double, _ c: Double) -> Double
}

class OrdinaryDiscriminantStrategy : DiscriminantStrategy
{
  func calculateDiscriminant(_ a: Double, _ b: Double, _ c: Double) -> Double
  {
    return b*b - 4*a*c
  }
}

class RealDiscriminantStrategy : DiscriminantStrategy
{
  func calculateDiscriminant(_ a: Double, _ b: Double, _ c: Double) -> Double
  {
    let result = b*b - 4*a*c
    return (result >= 0) ? result : Double.nan
  }
}

class QuadraticEquationSolver
{
  private let strategy: DiscriminantStrategy

  init(_ strategy: DiscriminantStrategy)
  {
    self.strategy = strategy
  }

  func solve(_ a: Double, _ b: Double, _ c: Double) -> (Double, Double)
  {
    let disc = strategy.calculateDiscriminant(a, b, c)
    let rootDisc = sqrt(disc)
    return ((-b + rootDisc) / (2*a), (-b - rootDisc) / (2*a))
  }
}

// ===

class UMBaseTestCase : XCTestCase {}

class Evaluate: UMBaseTestCase
{
  func positiveTestOrdinaryStrategy()
  {
    let strategy = OrdinaryDiscriminantStrategy()
    let solver = QuadraticEquationSolver(strategy)
    let (x1, x2) = solver.solve(1, 10, 16)

    XCTAssertEqual(-2, x1)
    XCTAssertEqual(-8, x2)
  }

  func negativeTestRealStrategy()
  {
    let strategy = RealDiscriminantStrategy()
    let solver = QuadraticEquationSolver(strategy)
    let (x1, x2) = solver.solve(1, 4, 5)

    XCTAssert(x1.isNaN)
    XCTAssert(x2.isNaN)
  }
}

extension Evaluate
{
  static var allTests : [(String, (Evaluate) -> () throws -> Void)]
  {
    return [
      ("positiveTestOrdinaryStrategy", positiveTestOrdinaryStrategy),
      ("negativeTestRealStrategy", negativeTestRealStrategy)
    ]
  }
}

func main()
{
  XCTMain([testCase(Evaluate.allTests)])
}

main()